### PR TITLE
nanomq: 0.20.8 → 0.21.8

### PR DESCRIPTION
**Pull request by sikmir**
## Description of changes
https://github.com/nanomq/nanomq/releases

## Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 284345` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nanomq</li>
  </ul>
</details>

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 284345` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nanomq</li>
  </ul>
</details>

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 284345` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nanomq</li>
  </ul>
</details>

**Comment by sikmir** [comment]
I have no idea why local builds is OK on linux/darwin, but CI fails.

File: pkgs/by-name/na/nanomq/package.nix [hunk 1]
```diff
@@ -62,16 +62,16 @@ in stdenv.mkDerivation (finalAttrs: {
...
+    (lib.cmakeBool "ENABLE_RULE_ENGINE" true)
+    (lib.cmakeBool "NNG_ENABLE_SQLITE" true)
+    (lib.cmakeBool "NNG_ENABLE_TLS" true)
   ];
 
-  env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isClang "-Wno-return-type -Wno-implicit-function-declaration -Wno-error=implicit-int";
+  env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isClang "-Wno-error=int-conversion";
 
   # disabled by default - not 100% reliable and making nanomq depend on
   # mosquitto would annoy people
```
Comment (SuperSandro2000) [inline pkgs/by-name/na/nanomq/package.nix:77]: nanaomq is not executing tests and they are failing for ofborg.

Comment (sikmir) [reply pkgs/by-name/na/nanomq/package.nix:77]: Do you mean ofborg tries to execute `withInstallChecks`?

Comment (SuperSandro2000) [reply pkgs/by-name/na/nanomq/package.nix:77]: I am not sure exactly what ofborg is doing. You can take a look at logs by clicking details below.

Comment (SuperSandro2000) [reply pkgs/by-name/na/nanomq/package.nix:77]:
```
Running phase: fixupPhase
shrinking RPATHs of ELF executables and libraries in /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8
shrinking /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8/bin/nanomq_cli
shrinking /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8/bin/nanomq
shrinking /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8/bin/nngcat
checking for references to /build/ in /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8...
patching script interpreter paths in /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8
stripping (with command strip and flags -S -p) in  /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8/lib /nix/store/p51fx920kwbhz1wfd82y3d716j16q186-nanomq-0.21.8/bin
Running phase: installCheckPhase
mqtt v311 test start
V4/V5 test passed!
clean session test passed!
Retain test passed!
mqtt v311 test end
mqtt v5 test start
User property test passed!
Shared subscription test passed!
Sub client did not receive message * 10
mosquitto_sub -h 127.0.0.1 -p 1883 -t 'topic'
mosquitto_pub -h 127.0.0.1 -p 1883 -t topic -V 5 -m message -D Publish topic-alias 10 -d --repeat 10
Topic alias test failed!
mqtt v5 test failed
2024-05-01 09:20:13 [4593] DEBUG /build/source/nanomq/pub_handler.c:1645: payload: [message], len = 7
2024-05-01 09:20:13 [4593] DEBUG /build/source/nng/src/supplemental/nanolib/mqtt_db.c:734: Searching client: topic
2024-05-01 09:20:13 [4593] DEBUG /build/source/nng/src/supplemental/nanolib/mqtt_db.c:737: Searching client: topic
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1190: pipe_info size: [2]
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1610: property len: 3
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1633: used pos: [6]
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1645: payload: [message], len = 7
2024-05-01 09:20:13 [4602] DEBUG /build/source/nng/src/supplemental/nanolib/mqtt_db.c:734: Searching client: topic
2024-05-01 09:20:13 [4602] DEBUG /build/source/nng/src/supplemental/nanolib/mqtt_db.c:737: Searching client: topic
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1190: pipe_info size: [2]
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/apps/broker.c:548: WAIT ^^^^ ctx83 ^^^^
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1351: start encode message
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1388: after topic and id len in msg already [7]
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1448: after payload len in msg already [18]
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1458: header len [2] remain len [18]
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1510: end encode message
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/apps/broker.c:548: WAIT ^^^^ ctx78 ^^^^
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1351: start encode message
2024-05-01 09:20:13 [4600] DEBUG /build/source/nanomq/pub_handler.c:1645: payload: [message], len = 7
2024-05-01 09:20:13 [4600] DEBUG /build/source/nng/src/supplemental/nanolib/mqtt_db.c:734: Searching client: topic
2024-05-01 09:20:13 [4600] DEBUG /build/source/nng/src/supplemental/nanolib/mqtt_db.c:737: Searching client: topic
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1388: after topic and id len in msg already [7]
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1448: after payload len in msg already [18]
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1458: header len [2] remain len [18]
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1510: end encode message
2024-05-01 09:20:13 [4591] DEBUG /build/source/nanomq/pub_handler.c:1510: end encode message
2024-05-01 09:20:13 [4600] DEBUG /build/source/nanomq/pub_handler.c:1190: pipe_info size: [2]
2024-05-01 09:20:13 [4591] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:452: pipe 539152686 occupied! resending in cb!
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/apps/broker.c:548: WAIT ^^^^ ctx80 ^^^^
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1351: start encode message
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1388: after topic and id len in msg already [7]
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1448: after payload len in msg already [18]
2024-05-01 09:20:13 [4593] DEBUG /build/source/nanomq/pub_handler.c:1190: pipe_info size: [2]
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1458: header len [2] remain len [18]
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1510: end encode message
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/apps/broker.c:548: WAIT ^^^^ ctx84 ^^^^
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1351: start encode message
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1388: after topic and id len in msg already [7]
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1448: after payload len in msg already [18]
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1458: header len [2] remain len [18]
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1510: end encode message
2024-05-01 09:20:13 [4602] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:430: Drop msg due to qos == 0
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1286: free topic
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1293: free properties
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1302: free payload
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:1308: free pub_packet
2024-05-01 09:20:13 [4602] DEBUG /build/source/nanomq/pub_handler.c:88: pub_handler: init pipe_info
2024-05-01 09:20:13 [4598] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:430: Drop msg due to qos == 0
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1286: free topic
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1293: free properties
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1302: free payload
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:1308: free pub_packet
2024-05-01 09:20:13 [4598] DEBUG /build/source/nanomq/pub_handler.c:88: pub_handler: init pipe_info
2024-05-01 09:20:13 [4591] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:430: Drop msg due to qos == 0
2024-05-01 09:20:13 [4591] DEBUG /build/source/nanomq/pub_handler.c:1286: free topic
2024-05-01 09:20:13 [4591] DEBUG /build/source/nanomq/pub_handler.c:1293: free properties
2024-05-01 09:20:13 [4591] DEBUG /build/source/nanomq/pub_handler.c:1302: free payload
2024-05-01 09:20:13 [4591] DEBUG /build/source/nanomq/pub_handler.c:1308: free pub_packet
2024-05-01 09:20:13 [4591] DEBUG /build/source/nanomq/pub_handler.c:88: pub_handler: init pipe_info
2024-05-01 09:20:13 [4601] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:452: pipe 539152686 occupied! resending in cb!
2024-05-01 09:20:13 [4606] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:430: Drop msg due to qos == 0
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1286: free topic
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1293: free properties
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1302: free payload
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:1308: free pub_packet
2024-05-01 09:20:13 [4606] DEBUG /build/source/nanomq/pub_handler.c:88: pub_handler: init pipe_info
2024-05-01 09:20:13 [4601] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:430: Drop msg due to qos == 0
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1286: free topic
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1293: free properties
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1302: free payload
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:1308: free pub_packet
2024-05-01 09:20:13 [4601] DEBUG /build/source/nanomq/pub_handler.c:88: pub_handler: init pipe_info
2024-05-01 09:20:13 [4604] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:235: check pipe keepalive interval 60 backoff -90000.000000, ka 0
2024-05-01 09:20:14 [4595] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:235: check pipe keepalive interval 60 backoff -89000.000000, ka 1
2024-05-01 09:20:15 [4596] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:235: check pipe keepalive interval 60 backoff -88000.000000, ka 2
2024-05-01 09:20:16 [4600] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:235: check pipe keepalive interval 60 backoff -87000.000000, ka 3
2024-05-01 09:20:17 [4592] DEBUG /build/source/nng/src/sp/protocol/mqtt/nmq_mqtt.c:235: check pipe keepalive interval 60 backoff -86000.000000, ka 4
2024-05-01 09:20:18 [4585] ERROR /build/source/nanomq/apps/broker.c:104: signal signumber: 15 received!
Traceback (most recent call last):
  File "/build/source/.github/scripts/test.py", line 55, in <module>
    raise AssertionError
AssertionError
error: builder for '/nix/store/ql2biqwv56k0cc78il7zn097axdni33f-nanomq-0.21.8.drv' failed with exit code 1;
```

**Comment by risicle** [comment]
0.21.8 addresses CVE-2024-25767

A "pre-release" 0.21.9 also addresses CVE-2024-31040 & CVE-2024-31041, but it's unclear how they treat "pre-releases". Might be better to add a patch for those?

**Comment by risicle** [comment]
I think the tests might need paho-mqtt 2.x which there are attempts to bump to @ #288219

**Comment by risicle** [comment]
It does actually look like websocket support might be properly broken - even given an updated paho-mqtt , the websocket functional tests give "connection refused" for everything.

We're going to have to decide whether we want to bump to a possibly partially broken nanomq or mark this with `knownVunerabilties`

**Pull request closed without merging**